### PR TITLE
[OC] Add compiler's role and memory/caches

### PR DESCRIPTION
**Pull request by constancadcunha**

### CODE REVIEW

File: content/oc/0003-papel-compilador.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
+---
+title: Papel de um compilador
+description: >-
```
Comment (diogotcorreia) [inline content/oc/0003-papel-compilador.md:3]: Missing description

File: content/oc/0003-papel-compilador.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+path: /oc/papel-compilador
+type: content
+---
+
+# Papel de um computador
```
Comment (diogotcorreia) [inline content/oc/0003-papel-compilador.md:9]: Title does not match page name

File: content/oc/0003-papel-compilador.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+Agora que já temos o offset, podemos calcular o nosso número de bloco! Através de 300 $mod$ 256, que nos dá 44, o nosso [_index_](color:pink).
+
+Podemos por isso concluir que blocos maiores [deviam reduzir](color:purple) o _miss rate_ devido à localidade espacial, mas numa cache de tamanho fixo, [blocos maiores implica menos blocos](color:purple), o que signfica que mais competição implica um _miss rate_ mais elevado e portanto, blocos maiores implicam poluição. Por isso, uma maior penalidade de _miss_ pode dar _override_ ao benefício de um reduzido _miss rate_ e um recomeço mais cedo pode ajudar!
+
+### Miss Rate vs Block Size vs Cache Size
+
+![Comparação](./assets/0003-comparacao.png#dark=3)
+
+Como podemos ver no gráfico acima, o _miss rate_ sobe quando o tamanho do bloco se uma fração mais significativa do tamanho da cache visto que o número de bloco que pode ser contido no mesmo tamanho de cache é mais pequeno, o que significa que [aumenta a capacidade de falhas](color:pink).
```
Comment (D-Card) [inline content/oc/0003-papel-compilador.md:190]: Missing words?

File: content/oc/0003-papel-compilador.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+:::warning[_Pitfalls_]
+
+Sequências de palavras [não são](color:red) endereços sequenciados, incrementamos sempre a 4 e não a 1! Manter o [ponteiro numa variável automática](color:red) a seguir a um procedimento retornar, por exemplo passar um ponteiro através de um argumento, visto que o ponteiro se torna inválido quando o stack é libertado.
+
+:::
+
+## Memória
+
+Antes de começarmos temos que relembrar a [diferença entre o CPU e RAM](color:purple). O RAM é a memória, o processador faz os cálculos/computações. Basicamente se tivermos um carpinteiro atrabalhar num banco e tivermos uma caixa com ferramentas, em qualquer momento só um certo número de ferramentas é que está a ser usado para consertar o banco, estas ferramentas encontram-se em cima do banco, o resto está na caixa. Usar só as ferramentas que já estão em cima do banco é mais rápido do que tirar uma ferramenta da caixa e usá-la. Assim, aplicando esta analogia a um computador, o [carpinteiro é o CPU](color:red), as [ferramentas que estão no banco são a RAM](color:pink) e [as ferramentas que estão na caixa são o disco rígido](color:purple).
```
Comment (D-Card) [inline content/oc/0003-papel-compilador.md:78]:
```suggestion
Antes de falar de memória é importante distinguir a [diferença entre o CPU e RAM](color:purple). A RAM é a memória primária enquanto que o processador faz os cálculos/computações. Pegando num exemplo, se tivermos um carpinteiro a arranjar uma cadeira e este tiver uma caixa com ferramentas, num dado momento só um certo número de ferramentas é que está a ser usado para consertar a cadeira. Estas ferramentas encontrar-se-ão mais próximo do carpinteiro, por exemplo no seu cinto, enquanto o resto estará na caixa. Usar as ferramentas que estão no seu cinto é mais rápido do que tirar uma ferramenta da caixa e usá-la. Se o carpinteiro for esperto, irá ter no seu cinto as ferramentas que precisa mais frequentemente para a tarefa que está a fazer no momento e deixará na caixa as ferramentas menos utilizadas. Assim, aplicando esta analogia a um computador, o [carpinteiro é o CPU](color:red), as [ferramentas que estão no cinto são a RAM](color:pink) e [as ferramentas que estão na caixa são o disco rígido](color:purple).
```

**Review by D-Card — CHANGES_REQUESTED** [review]

**Review by D-Card — CHANGES_REQUESTED** [review]

**Review by diogotcorreia — APPROVED** [review]

**Pull request merged**